### ISSUE TITLE
Pull request testing should set -Werror

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#3178]
@vbrunini asks whether pull request testing could set `-Werror`, so as to avoid issues like #3177.

@trilinos/framework @khpierson 

## Expectations

Trilinos -- at least the library, not necessarily tests and examples -- should build warning-free.

## Current Behavior

See #3177.  There is an issue that it's impossible to fix warnings in some packages.

## Motivation and Context

Sierra builds with warnings as errors, so they want Trilinos to build warning-free.

## Possible Solution

Exclude legacy packages like ML.  Fix warnings.  Add `-Werror` to at least one PR build.

## Related Issues
* Related to #3177

**Comment by bartlettroscoe** [comment]
We have to clean up the warnings in the PR builds before you can set `-Werror`.  However, we could do this package-by-package.  There is a way to do that with TriBITS I think (and if not, we can make sure there is).

**Comment by bartlettroscoe** [comment]
CC: @jwillenbring, @srajama1, @kddevin, @rppawlo, @mperego (The Trilinos Product Area Leads)

@trilinos/framework, can we make this the official issue for cleaning up the warnings in the PR builds, one package at a time and setting `-Werror` as we have been discussing for the last several months?  (I could not find any other Trilinos GitHub issue that was tracking this effort.)

Also, can we please add the following warning flags to the GCC and Intel Trilinos PR builds before we start setting -Werror?  These were specifically requested by the ATDM Gemma code and they really need to have zero Trilinos warnings.  They have indicated that this is a major impediment to using Trilinos and to improving their own software quality.


1. For gcc/clang:
     -Wall
     -Wextra

2. For intel compilers (icc):
     -Wall  
     -Warray-bounds
     -Wchar-subscripts  
     -Wcomment   
     -Wenum-compare 
     -Wformat 
     -Wuninitialized 
     -Wmaybe-uninitialized 
     -Wmain 
     -Wnarrowing 
     -Wnonnull 
     -Wparentheses 
     -Wpointer-sign 
     -Wreorder 
     -Wreturn-type 
     -Wsign-compare 
     -Wsequence-point 
     -Wtrigraphs 
     -Wunused-function 
     -Wunused-but-set-variable 
     -Wunused-variable 
     -Wwrite-strings

**Comment by srajama1** [comment]
@bartlettroscoe Thanks for taking the lead on this. I like the idea of starting with minimal set of Trilinos packages and keep adding more to that list.

**Comment by khpierson** [comment]
FYI, I've investigated using -Wextra in sierra a few times and noted that we have tens of thousands of extra warnings coming from the compiler with that setting enabled. Sierra does not build with -Wextra by default, just -Wall and -Werror for gcc as far as I know.

Of course, if you can achieve clean builds with -Wextra, that's great!

**Comment by jmgate** [comment]
I'd also throw in `-Wpedantic`, but then we have a much longer way to go before we're clean.

**Comment by bartlettroscoe** [comment]
@khpierson said:

> Of course, if you can achieve clean builds with -Wextra, that's great!

Trilinos (or at least Trilinos header files) has to build clean with the union of all the warning options that Trilinos customers are using.  So if an important (paying) customer needs Trilinos to build clean with `-Wextra`, we need to somehow make that happen.  That is just my opinion but the SQA for Trilinos needs to be as high or higher than the customer's own SQA practices.

**Comment by prwolfe** [comment]
@khpierson - even that is a moving target as you are aware. The warnings included in the gcc version of -Wall has been greatly reduced over the years. Still, a common target of whatever sort would be useful.

**Comment by khpierson** [comment]
Who actually prioritizes fixing/addressing these issues? 
I would like to determine the best way to engage from the SIERRA perspective.

**Comment by srajama1** [comment]
@khpierson The best way is to engage the product leads (all copied above). This kind of issue usually is in the scope of @trilinos/framework , so @jwillenbring would be the right person. However, in this specific case @bartlettroscoe has taken the lead because it came from an ATDM application.

**Comment by khpierson** [comment]
Is there a product lead meeting? The reason I'm asking is that I'm interested in attending once in awhile to better understand how SIERRA can engage. If that's not the correct forum, I would like to find somewhere to figure this out. Thanks!

**Comment by mhoemmen** [comment]
There used to be regular Sierra / Trilinos / other apps phone meetings, but they died.  Mike Heroux was leading those, but he got busy with other things.

**Comment by khpierson** [comment]
Is there a single point of contact? or is Siva correct in saying there is a counsel in charge?

**Comment by jwillenbring** [comment]
After discussions with @trilinos/framework @bartlettroscoe, and @fryeguy52 the current plan is to attack this very incrementally. The plan can be divided into three primary parts:

1) Add SIERRA's warnings as errors flags to the GCC 7.2 PR build. These flags, as reported by @rrdrake are:
`
-Wall -Wno-clobbered -Wno-vla -Wno-pragmas -Wno-unknown-pragmas -Wno-unused-local-typedefs -Wno-literal-suffix -Wno-deprecated-declarations -Wno-misleading-indentation -Wno-int-in-bool-context -Wno-maybe-uninitialized -Wno-nonnull-compare -Wno-address -Wno-inline -Werror
`

With some pointers from @bartlettroscoe we were able to set this up at a package level granularity. 

This effort is already in progress. These flags have been added to the GCC 7.2 PR (and dev->master) build for [Teuchos](https://github.com/trilinos/Trilinos/pull/4374) and [Kokkos](https://github.com/trilinos/Trilinos/pull/4384).

Next step is to do a broader evaluation of how many warnings need to be cleaned up and work through more packages, protecting the introduction of new warnings at the package level.

2) Add Gemma flags to the GCC 7.2 build. (This is "just" -wextra at this point). This can be started before finishing 1). Once packages have the SIERRA flags applied, we can try -wextra at any time and file those tickets also and add the flag to the build when all issues are resolved at a package level.

3) Add Gemma flags to the icc build. We need to baseline where we are at with these flags and look at incrementally adding them. ICC has not been a target for warnings as errors efforts previously, so this might take considerable effort. We should potentially also wait until Intel 18 is available for PR testing instead of starting with 17.

Previous efforts to clean up warnings have never been fully completed due to warnings in a small number of (mostly older) packages. For this effort, we need to be able to add these flags, so where the warnings cannot be cleaned up, we will need to use a strategy to minimally silence the warnings. It might be best to do this in if blocks so we can enable the warnings on-demand. We will flesh out these details as the effort moves forward.

**Comment by jmgate** [comment]
> It might be best to do this in if blocks so we can enable the warnings on-demand.

You probably want to touch base with @CamelliaDPG to talk through what EMPIRE's done along these lines lately.

**Comment by CamelliaDPG** [comment]
Our basic approach to cleaning up warnings in EMPIRE has been this:
1. Issue Trilinos PRs for warnings generated by Trilinos.  Most, but not all, of these have been addressed.
2. For particular includes that cause warnings, create a "quiet_{HEADER_NAME}.hpp" in our project.
3. Replace all those includes with the quiet variants.

For us, most of the compiler-specific warnings come from clang (we do have a number that are nvcc specific, but we have not yet developed a way to silence these yet).  The quiet headers are therefore of the form

```
#if defined (__clang__)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wformat-pedantic"
#pragma GCC diagnostic ignored "-Wunused-private-field"
#endif

#include "the_actual_header.hpp"

#if defined (__clang__)
#pragma GCC diagnostic pop
#endif
```

We've added headers along these lines for Panzer's STK-including headers, Tpetra MatrixMarket, and H5hut.h.

**Comment by mhoemmen** [comment]
@CamelliaDPG Please don't forget to file issues for warnings, and to remove the pragmas once the issues are fixed -- thanks!

**Comment by alanw0** [comment]
We finally have some work in progress on the STK team to remove boost from stk-topology, which is a major source of warnings for Nvidia builds. In the sierra eco-system (where stk is developed) we have pretty strong warning flags on these days, i.e., -Wextra -Wpedantic -Wshadow for the gcc build. We're actively working towards being warning free.

**Comment by ZUUL42** [comment]
I've been able to successfully turn on Werror for @trilinos/kokkos-kernels, @trilinos/sacado, @trilinos/shards, @trilinos/phalanx, & @trilinos/panzer. PR #4405
For the rest I'm looking to give teams some feedback as to what is needed to _git_ this in.
I can start going through and posting CDash reports to give teams package-specific results.

Here are [the results](https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&parentid=4544770) from a build for @trilinos/tpetra. Just looks like a number of unused-x warnings that need to be cleaned up.
I'll post more results as they finish up.

Let me know if seeing build results in CDash helps, doesn't, or if there's anything else that might be useful.

**Comment by mhoemmen** [comment]
@trilinos/tpetra I'll number the Tpetra 12 warnings in the order shown here: https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4544817

  - 2 (Export): already fixed; code was removed.
  - 3 (Import): ditto

I'm working on the rest.

**Comment by ZUUL42** [comment]
Here are the results from some test builds I let run last night with Werror turned on:
@trilinos/tpetra [CDash](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4584015)
@trilinos/zoltan2 [CDash](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4584069)
@trilinos/epetra [CDash](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4584123)

**Comment by ZUUL42** [comment]
@trilinos/piro appears to only have one error with Werror turned on
[CDash](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4589529)

**Comment by ZUUL42** [comment]
[Results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4589916) for @trilinos/belos

**Comment by ZUUL42** [comment]
[Results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4590377) for @trilinos/ifpack2 
[Results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4595659) for @trilinos/intrepid2 
[Results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4595713) for @trilinos/anasazi 
[Results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4595605) for @trilinos/rol

**Comment by kddevin** [comment]
Zoltan2's warnings are fixed by #4436.

**Comment by ZUUL42** [comment]
@kddevin, thanks. I'll get a PR in to turn Werror on for Zoltan2 in the build shortly.

**Comment by ZUUL42** [comment]
[Updated Werror results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4599045) for @trilinos/tpetra

**Comment by mhoemmen** [comment]
I'm confused about the first warning.  It claims that `iallreduceIntRaw` is defined but not used.  However, I can see that it's used here, in a way not protected by a macro:

https://github.com/trilinos/Trilinos/blob/5f47fa046b44ffb0273ac1968773d67d396950e5/packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp#L8234

For the other warnings, I'm in all-day meetings all week and likely won't have time to fix it until next week.  Anybody else want to take it?

**Comment by ZUUL42** [comment]
I noticed that one seemed odd as well. Thought it might either be another issue upstream presenting as an unused-function, but don't know the code base well enough to effectively dig into it quickly.
If it's found to be a totally false-positive, we can discuss how we should deal with it.

**Comment by ZUUL42** [comment]
Building from [branch ZUUL42/Werror_dfi](https://github.com/ZUUL42/Trilinos/tree/Werror_dfi) to trilinos/develop, here are the results for:
- [results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4623054) @trilinos/domi
- [results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4623057) @trilinos/fei 
- [results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4623061) @trilinos/ifpack 

Then building from [branch ZUUL42/Werror_ntt](https://github.com/ZUUL42/Trilinos/tree/Werror_ntt) to trilinos/develop, here are the results for:
- [results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4623125) @trilinos/nox 
- [results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4623145) @trilinos/tempus 
- [results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4623152) @trilinos/triutils 

This should be the last of the warnings->errors that prevent setting -Werror.

**Comment by ZUUL42** [comment]
Here are warnings that turn into errors that need to be fixed before we can set -Werror for the following packages:
- @trilinos/ml => [Results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4624516)
- @trilinos/muelu => [Results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4624462)
- @trilinos/seacas => [Results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4624408)
- @trilinos/stokhos => [Results](https://testing-vm.sandia.gov/cdash/viewBuildError.php?buildid=4624570)

**Comment by ZUUL42** [comment]
I've compiled ICC build results for all packages with Werror set individually.
The CDash results can be found [here](https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercombine=and&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=PR-Werror_icc_&field2=buildstarttime&compare2=84&value2=NOW).
Some results may be redundant for packages that don't yet have fixes in for GCC 7.2.0 Werror.
A few packages have hit the 50 Error limit. I'll be looking into a good way to break that down to give teams a full report. Otherwise, the first 50 errors can be worked and I will create a second build once the first set of fixes are in.

**Comment by etphipp** [comment]
Regarding one of the Sacado warnings/errors, I don't think the warning 'loop was not vectorized with "simd"' is one that should be turned into an error.  That just means the loop couldn't be vectorized, which there are a lot of valid reasons for, and it is going to be incredibly common with code that uses Kokkos.

**Comment by ZUUL42** [comment]
@etphipp, thanks. I'll look into that.

**Comment by mperego** [comment]
Most of Intrepid2 warnings are of the type: "variable xxx was declared but never referenced".
In particular for the following piece of code,
 `const ordinal_type
      C = outputFields.extent(0),
      F = outputFields.extent(1);
    
    using range_policy_type = Kokkos::Experimental::MDRangePolicy
        < ExecSpaceType, Kokkos::Experimental::Rank<2>, Kokkos::IndexType<ordinal_type> >;
    const range_policy_type policy( { 0, 0 },
                                    { C, F} );`

The variables `C`, `F` and  are not referenced when the execution space is serial. How should I fix this? I guess I can simply declare the `range policy` as
`const range_policy_type policy( { 0, 0 },
                                    { outputFields.extent(0),  outputFields.extent(1) } );`

But is it worth it?

**Comment by kyungjoo-kim** [comment]
@mperego What is the routine that complains about it ? What you mean is that md range policy with serial space does not use the C and F ?

**Comment by mperego** [comment]
@kyungjoo-kim,  look at the errors here https://testing-vm.sandia.gov/cdash/buildSummary.php?buildid=4724605

**Comment by kyungjoo-kim** [comment]
@mperego  I don't understand why the compiler complains. This does not make sense and **cannot** and **should not**  generate warnings. I remembered that you cleaned up all the warnings a while ago. This warning is non-sense to me and we cannot remove these warnings. 

@ndellingwood @ibaned Do you have an idea why md range policy generates this warning ?

**Comment by ibaned** [comment]
I've seen this warning pattern before, and I agree it looks like a compiler bug, since the variables are referenced in policy creation. I haven't found a better solution than inlining the variables as @mperego suggests. Perhaps trying a newer compiler to see if the compiler bug is fixed would be a better use of time?

**Comment by mperego** [comment]
@mhoemmen @bartlettroscoe @ZUUL42 how would you like to proceed? It's not too much trouble to inline the variables, but I don't think we should make the code less readable only to avoid a warning likely due to a compiler bug.

**Comment by bartlettroscoe** [comment]
@mperego asked:

> how would you like to proceed? It's not too much trouble to inline the variables, but I don't think we should make the code less readable only to avoid a warning likely due to a compiler bug.

The warnings just need to go away.  Either change to code to remove the warnings or put in a compiler warning disable around the piece of code generating the warning using preprocessor ifdefs and pragmas with push and pop.  See:

* https://codeyarns.com/2014/03/11/how-to-selectively-ignore-a-gcc-warning/
* https://stackoverflow.com/questions/3378560/how-to-disable-gcc-warnings-for-a-few-lines-of-code

But if there lots of code like this disbursed all over the code base, the preprocessor solution would be super ugly.  Or to make this easier you could put these around larger blocks of code.

As for code changes, perhaps use macros for these types of use cases so that in non-parallel builds the variables never even get declared?  I would have to see all of the different use cases to know if this is feasible but such macros might not be too bad.  Someones you can define the macro so that it uses tricks like `(void)<var_name>` to make unused warnings go away.  Just have to experiment some.

How many unique warnings like this are we talking about here?

**Comment by mperego** [comment]
@bartlettroscoe It's only an handful of warnings. I fixed it (or I hope I have) by inlining some code.

**Comment by bartlettroscoe** [comment]
@mperego said:

> @bartlettroscoe It's only an handful of warnings. I fixed it (or I hope I have) by inlining some code.

@mperego, can you not reproduce these warnings locally on your own machine?  That should be easy to do on any COE RHEL6 machine that has the SEMS env.

**Comment by mperego** [comment]
@bartlettroscoe I have RHEL7. I could do that on a cee machine, but I'd rather have it checked by the automated tests. A little waste of computational resources if the problem does not get fixed, but it would save me some time.

**Comment by bartlettroscoe** [comment]
@mperego said:

> @bartlettroscoe I have RHEL7. I could do that on a cee machine, but I'd rather have it checked by the automated tests. A little waste of computational resources if the problem does not get fixed, but it would save me some time.

@mperego, those SEMS modules work on RHEL7 as well.  Sound be like falling off log to reproduce locally.

The PR system is overburdened as it is right now.

**Comment by mperego** [comment]
OK, how can I set the Werror ONLY for the Intrepid2 package? I've trouble today accessing the https://testing-vm.sandia.gov/cdash/ (insecure connection)

**Comment by lucbv** [comment]
Adding this should add warnings as error in Intrepid2:
```
-D   Intrepid2_CXX_FLAGS="${CXX_WARNINGS} -Werror"
```

**Comment by lucbv** [comment]
You probably want to set `CXX_WARNINGS` appropriately for the exact set of warnings to be tested...

**Comment by bartlettroscoe** [comment]
@mperego said:

>  I've trouble today accessing the https://testing-vm.sandia.gov/cdash/ (insecure connection)

Replace 'testing-vm' with 'testing'.  Not sure why testing-vm is doing that but the real machine is now testing.sandia.gov.

**Comment by bartlettroscoe** [comment]
@trilinos/framework, it would seem to be strongly recommend that after #4841 is merged that `-Werror` be set globally for that GCC 7.2.0 build and take out the `-Werror` for each individual package.  That will ensure that when new packages get added, they will not be allowed to have any warnings either.   Make sense?

**Comment by ZUUL42** [comment]
@bartlettroscoe that's the plan. First we still have 12 packages that do not pass when `-Werror` is set, some small, some extensive.
Once those fixes get it, I'll set `-Werror` globally.

**Comment by ZUUL42** [comment]
These are the packages with warnings turning into errors under `-Werror` remaining in the GCC 7.2.0 build.
Once these are cleared up we can set `-Werror` globally for GCC 7.2.0.

[All builds at a glance on CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=Werror_&field2=buildstarttime&compare2=83&value2=2019/04/09)

@trilinos/amesos - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4863709)
@trilinos/anasazi - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4863763) - #4475
@trilinos/aztecoo - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4863817)
@trilinos/belos - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4863871) - #4473
​@trilinos/epetra - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4863925) - #4468
@trilinos/fei - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4863979) - #4507
@trilinos/isorropia - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4864033)
@trilinos/pamgen - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4864392)
@trilinos/rol - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4864141) - #4469
@trilinos/tpetra - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4864204)
@trilinos/triutils - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4864284) - #4511
@trilinos/zoltan - [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=4864338)

**Comment by mhoemmen** [comment]
Regarding @trilinos/tpetra, we've talked before about how this one is spurious:
```
In file included from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_7.2.0/Trilinos/packages/tpetra/core/src/Tpetra_CrsMatrix_convert.cpp:50:0: /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_7.2.0/Trilinos/packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp:89:3: error: ‘std::shared_ptr<Tpetra::Details::CommRequest> Tpetra::{anonymous}::iallreduceIntRaw(const int&, int&, Teuchos::EReductionType, const Teuchos::Comm<int>&)’ defined but not used [-Werror=unused-function]    iallreduceIntRaw (const int& localValue,    ^~~~~~~~~~~~~~~~ cc1plus: all warnings being treated as errors
--
```
We've worked around compiler bugs before, though ;-) .

**Comment by mhoemmen** [comment]
PR #4862 addresses all Tpetra build warnings but the spurious one mentioned in my previous comment. 
 Working around that one will take a bit more effort and really should be a in a separate PR.

**Comment by bartlettroscoe** [comment]
@mhoemmen said:

> PR #4862 addresses all Tpetra build warnings but the spurious one mentioned in my previous comment.
Working around that one will take a bit more effort and really should be a in a separate PR.

@mhoemmen, can you just use ifdefs and pragmas to disable just this warning surgically?  See links for how to this [above](https://github.com/trilinos/Trilinos/issues/3178#issuecomment-475240344).

**Comment by mhoemmen** [comment]
@bartlettroscoe wrote:
> ... can you just use ifdefs and pragmas to disable just this warning surgically? See links for how to this [above](https://github.com/trilinos/Trilinos/issues/3178#issuecomment-475240344).

I don't think the DevOps person in question feels like they have the authority to disable warnings.  I'll do my best to fix them, but I'm not sure if it's possible.

**Comment by ZUUL42** [comment]
`-Werror` has now been set globally for the GCC 7.2.0 PR build.

**Comment by bartlettroscoe** [comment]
@mhoemmen, @william76, @jmgate, @trilinos/framework 

FYI: Potential nice approach for managing targeted disabling of warnings on different compilers is given in:

* https://www.fluentcpp.com/2019/08/30/how-to-disable-a-warning-in-cpp/

This looks very good.

**Comment by mhoemmen** [comment]
@bartlettroscoe Thanks!  I'm not sure if we want to take this approach for CUDA warnings, though.  There are too many spurious warnings for us to disable each one, but I'm not sure if it's wise to disable them for an entire header file.

**Comment by bartlettroscoe** [comment]
@mhoemmen said:

>  I'm not sure if it's wise to disable them for an entire header file.

With the approach in:

* https://www.fluentcpp.com/2019/08/30/how-to-disable-a-warning-in-cpp/

I think you can disable a warning for single line of code if you want.

**Comment by etphipp** [comment]
SPARC is already doing this a lot to disable specific warnings when they pull in specific Trilinos headers.

**Comment by ibaned** [comment]
I think for "normal" CPU compiler warnings @bartlettroscoe 's approach is the right one when the code really has to be the way it is and can't be refactored to avoid the warning. The CUDA compiler emits warnings that are immune to this mechanism, however, so need slightly different workarounds.

**Comment by jwillenbring** [comment]
This has been done for some time.
